diff --git a/clang/lib/Basic/Targets.cpp b/clang/lib/Basic/Targets.cpp
index 8c01cfca8ccb..6e854190a62a 100644
--- a/clang/lib/Basic/Targets.cpp
+++ b/clang/lib/Basic/Targets.cpp
@@ -746,6 +746,9 @@ std::unique_ptr<TargetInfo> AllocateTarget(const llvm::Triple &Triple,
     case llvm::Triple::WASIp3:
       return std::make_unique<WASIP3TargetInfo<WebAssembly32TargetInfo>>(Triple,
                                                                          Opts);
+    case llvm::Triple::WasmWebP2:
+      return std::make_unique<WasmWebP2TargetInfo<WebAssembly32TargetInfo>>(
+          Triple, Opts);
     case llvm::Triple::Emscripten:
       return std::make_unique<EmscriptenTargetInfo<WebAssembly32TargetInfo>>(
           Triple, Opts);
diff --git a/clang/lib/Basic/Targets/OSTargets.h b/clang/lib/Basic/Targets/OSTargets.h
index 7c4666606d89..41cc013cf2bc 100644
--- a/clang/lib/Basic/Targets/OSTargets.h
+++ b/clang/lib/Basic/Targets/OSTargets.h
@@ -988,6 +988,20 @@ public:
   using WebAssemblyOSTargetInfo<Target>::WebAssemblyOSTargetInfo;
 };
 
+// WasmWebp2 target
+template <typename Target>
+class LLVM_LIBRARY_VISIBILITY WasmWebP2TargetInfo
+    : public WebAssemblyOSTargetInfo<Target> {
+  void getOSDefines(const LangOptions &Opts, const llvm::Triple &Triple,
+                    MacroBuilder &Builder) const final {
+    WebAssemblyOSTargetInfo<Target>::getOSDefines(Opts, Triple, Builder);
+    Builder.defineMacro("__webp2__");
+  }
+
+public:
+  using WebAssemblyOSTargetInfo<Target>::WebAssemblyOSTargetInfo;
+};
+
 // WALI target
 template <typename Target>
 class LLVM_LIBRARY_VISIBILITY WALITargetInfo
diff --git a/clang/lib/Driver/ToolChains/WebAssembly.cpp b/clang/lib/Driver/ToolChains/WebAssembly.cpp
index 7dabefbee7f7..32fd46ef7d16 100644
--- a/clang/lib/Driver/ToolChains/WebAssembly.cpp
+++ b/clang/lib/Driver/ToolChains/WebAssembly.cpp
@@ -72,8 +72,9 @@ std::string wasm::Linker::getLinkerPath(const ArgList &Args) const {
 static bool TargetBuildsComponents(const llvm::Triple &TargetTriple) {
   // WASIp2 and above are all based on components, so test for WASI but exclude
   // the original `wasi` target in addition to the `wasip1` name.
-  return TargetTriple.isOSWASI() && TargetTriple.getOSName() != "wasip1" &&
-         TargetTriple.getOSName() != "wasi";
+  return (TargetTriple.isOSWASI() && TargetTriple.getOSName() != "wasip1" &&
+          TargetTriple.getOSName() != "wasi") ||
+         TargetTriple.isOSWasmWeb();
 }
 
 static bool WantsPthread(const llvm::Triple &Triple, const ArgList &Args) {
@@ -139,6 +140,11 @@ void wasm::Linker::ConstructJob(Compilation &C, const JobAction &JA,
   const char *Crt1;
   const char *Entry = nullptr;
 
+  // Web-specific builds default to the reactor exec model.
+  if (ToolChain.getTriple().isOSWasmWeb()) {
+    IsCommand = false;
+  }
+
   // When -shared is specified, use the reactor exec model unless
   // specified otherwise.
   if (Args.hasArg(options::OPT_shared))
diff --git a/clang/lib/Lex/InitHeaderSearch.cpp b/clang/lib/Lex/InitHeaderSearch.cpp
index e894086b66e7..99bf0dba585d 100644
--- a/clang/lib/Lex/InitHeaderSearch.cpp
+++ b/clang/lib/Lex/InitHeaderSearch.cpp
@@ -234,6 +234,7 @@ bool InitHeaderSearch::ShouldAddDefaultIncludePaths(
   case llvm::Triple::WASIp1:
   case llvm::Triple::WASIp2:
   case llvm::Triple::WASIp3:
+  case llvm::Triple::WasmWebP2:
   case llvm::Triple::Win32:
   case llvm::Triple::ZOS:
     return false;
diff --git a/clang/test/Driver/wasm-toolchain.c b/clang/test/Driver/wasm-toolchain.c
index c02a102fab08..73acc14959dd 100644
--- a/clang/test/Driver/wasm-toolchain.c
+++ b/clang/test/Driver/wasm-toolchain.c
@@ -297,6 +297,35 @@
 // LINK_WASIP2_USE_WASMLD: "-cc1" {{.*}} "-o" "[[temp:[^"]*]]"
 // LINK_WASIP2_USE_WASMLD: wasm-ld{{.*}}" "-m" "wasm32" {{.*}} "[[temp]]" {{.*}}
 
+// Test that `wasm32-webp2` invokes the `wasm-component-ld` linker by default
+// instead of `wasm-ld`.
+
+// RUN: %clang -### -O2 --target=wasm32-webp2 %s --sysroot /foo 2>&1 \
+// RUN:   | FileCheck -check-prefix=LINK_WEBP2 %s
+// LINK_WEBP2: "-cc1" {{.*}} "-o" "[[temp:[^"]*]]"
+// LINK_WEBP2: wasm-component-ld{{.*}}" "-L/foo/lib/wasm32-webp2" "crt1-reactor.o" "--entry" "_initialize" "[[temp]]" "-lc" "{{.*[/\\]}}libclang_rt.builtins.a" "-o" "a.out"
+
+// RUN: %clang -### -O2 --target=wasm32-webp2 %s --sysroot /foo -mexec-model=command 2>&1 \
+// RUN:   | FileCheck -check-prefix=LINK_WEBP2_COMMAND %s
+// LINK_WEBP2_COMMAND: "-cc1" {{.*}} "-o" "[[temp:[^"]*]]"
+// LINK_WEBP2_COMMAND: wasm-component-ld{{.*}}" "-L/foo/lib/wasm32-webp2" "crt1.o" "[[temp]]"
+
+// Test that on `wasm32-webp2` the `wasm-component-ld` program is told where
+// to find `wasm-ld` by default.
+
+// RUN: %clang -### -O2 --target=wasm32-webp2 %s --sysroot /foo 2>&1 \
+// RUN:   | FileCheck -check-prefix=LINK_WEBP2_FIND_WASMLD %s
+// LINK_WEBP2_FIND_WASMLD: "-cc1" {{.*}} "-o" "[[temp:[^"]*]]"
+// LINK_WEBP2_FIND_WASMLD: wasm-component-ld{{.*}}" {{.*}} "--wasm-ld-path" "{{.*}}wasm-ld{{.*}}" {{.*}} "[[temp]]" {{.*}}
+
+// If `wasm32-webp2` is configured with `wasm-ld` as a linker then don't pass
+// the `--wasm-ld-path` flag.
+
+// RUN: %clang -### -O2 --target=wasm32-webp2 -fuse-ld=lld %s --sysroot /foo 2>&1 \
+// RUN:   | FileCheck -check-prefix=LINK_WEBP2_USE_WASMLD %s
+// LINK_WEBP2_USE_WASMLD: "-cc1" {{.*}} "-o" "[[temp:[^"]*]]"
+// LINK_WEBP2_USE_WASMLD: wasm-ld{{.*}}" "-m" "wasm32" {{.*}} "[[temp]]" {{.*}}
+
 // Basic `wasm32-linux-muslwali` compile-link test.
 
 // RUN: %clang -### --target=wasm32-linux-muslwali --sysroot=/foo %s 2>&1 \
diff --git a/clang/test/Preprocessor/init.c b/clang/test/Preprocessor/init.c
index 29a8d234ece4..09fa4874955c 100644
--- a/clang/test/Preprocessor/init.c
+++ b/clang/test/Preprocessor/init.c
@@ -1690,6 +1690,9 @@
 // RUN: %clang_cc1 -E -dM -ffreestanding -fgnuc-version=4.2.1 -triple=wasm64-wasip3 \
 // RUN:   < /dev/null \
 // RUN:   | FileCheck -match-full-lines -check-prefixes=WEBASSEMBLY,WEBASSEMBLY64,WASI,WASIP3 %s
+// RUN: %clang_cc1 -E -dM -ffreestanding -fgnuc-version=4.2.1 -triple=wasm32-webp2 \
+// RUN:   < /dev/null \
+// RUN:   | FileCheck -match-full-lines -check-prefixes=WEBASSEMBLY,WEBASSEMBLY32,WEBP2 %s
 // RUN: %clang_cc1 -E -dM -ffreestanding -fgnuc-version=4.2.1 -triple=wasm32-unknown-unknown -x c++ \
 // RUN:   < /dev/null \
 // RUN:   | FileCheck -match-full-lines -check-prefixes=WEBASSEMBLY-CXX %s
@@ -2134,6 +2137,7 @@
 // WEBASSEMBLY64-NEXT:#define __wasm64 1
 // WEBASSEMBLY64-NEXT:#define __wasm64__ 1
 // WEBASSEMBLY-NEXT:#define __wasm__ 1
+// WEBP2-NEXT:#define __webp2__ 1
 // EMSCRIPTEN:#define unix 1
 // WASI-NOT:#define unix 1
 // WEBASSEMBLY-CXX-NOT:_REENTRANT
diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index d819c4975b13..4571003253d2 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -336,6 +336,7 @@ public:
     WASIp1,     // WASI 0.1
     WASIp2,     // WASI 0.2
     WASIp3,     // WASI 0.3
+    WasmWebP2,  // Wasm component model + preview web bindings
     Emscripten,
     ShaderModel, // DirectX ShaderModel
     LiteOS,
@@ -838,6 +839,9 @@ public:
            getOS() == Triple::WASIp2 || getOS() == Triple::WASIp3;
   }
 
+  /// Tests whether the OS is the web.
+  bool isOSWasmWeb() const { return getOS() == Triple::WasmWebP2; }
+
   /// Tests whether the OS is Emscripten.
   bool isOSEmscripten() const { return getOS() == Triple::Emscripten; }
 
diff --git a/llvm/include/llvm/TargetParser/TripleName.def b/llvm/include/llvm/TargetParser/TripleName.def
index bbc03b318801..d86c64de3868 100644
--- a/llvm/include/llvm/TargetParser/TripleName.def
+++ b/llvm/include/llvm/TargetParser/TripleName.def
@@ -70,6 +70,7 @@ TRIPLE_OS(WASIp1, "wasip1")
 TRIPLE_OS(WASIp2, "wasip2")
 TRIPLE_OS(WASIp3, "wasip3")
 TRIPLE_OS(WASI, "wasi")
+TRIPLE_OS(WasmWebP2, "webp2")
 TRIPLE_OS(Emscripten, "emscripten")
 TRIPLE_OS(ShaderModel, "shadermodel")
 TRIPLE_OS(LiteOS, "liteos")
diff --git a/llvm/unittests/TargetParser/TripleTest.cpp b/llvm/unittests/TargetParser/TripleTest.cpp
index cfe3bc6abf72..d6b53413dab3 100644
--- a/llvm/unittests/TargetParser/TripleTest.cpp
+++ b/llvm/unittests/TargetParser/TripleTest.cpp
@@ -632,6 +632,12 @@ TEST(TripleTest, ParsedIDs) {
   EXPECT_EQ(Triple::WASIp3, T.getOS());
   EXPECT_EQ(Triple::UnknownEnvironment, T.getEnvironment());
 
+  T = Triple("wasm32-unknown-webp2");
+  EXPECT_EQ(Triple::wasm32, T.getArch());
+  EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
+  EXPECT_EQ(Triple::WasmWebP2, T.getOS());
+  EXPECT_EQ(Triple::UnknownEnvironment, T.getEnvironment());
+
   T = Triple("wasm64-unknown-unknown");
   EXPECT_EQ(Triple::wasm64, T.getArch());
   EXPECT_EQ(Triple::UnknownVendor, T.getVendor());
